{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import numpy as np\n",
    "# hide numpy log warning\n",
    "np.seterr(divide = 'ignore') \n",
    "import pandas as pd\n",
    "import warnings\n",
    "import unittest\n",
    "\n",
    "from biom import load_table, Table\n",
    "from biom.util import biom_open\n",
    "from scipy.signal import medfilt\n",
    "from scipy.special import softmax\n",
    "from skbio.stats.composition import closure\n",
    "from numpy.random import poisson, lognormal, normal, randint\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "# hide pandas Future/Deprecation Warning(s) for tutorial\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning) \n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def softmax_inv(x):\n",
    "    return np.log(x)\n",
    "\n",
    "class TestSoftInv(unittest.TestCase):\n",
    "\n",
    "    def setUp(self):\n",
    "        self.exp = closure(np.array([[100, 5, 20],\n",
    "                                     [200, 234, 14]]))\n",
    "        pass\n",
    "\n",
    "    def testinv(self):\n",
    "\n",
    "        res = softmax(softmax_inv(self.exp))\n",
    "        res *= 1 / res.sum(axis=1)[0]\n",
    "        try:\n",
    "            np.testing.assert_array_almost_equal(res, self.exp, 6)\n",
    "        except AssertionError:\n",
    "            print('arrays not equal')\n",
    "\n",
    "def simulate(X, depth=100, read_std=10.0,\n",
    "             pnormal=0.1, pheter=0.1,\n",
    "             pzero=0.1, mwindow=False):\n",
    "    \"\"\"\n",
    "    Poisson Log-Normal simulation\n",
    "    on real count data. The proceedure\n",
    "    follows the same steps as\n",
    "    Äijö et al. [1] but with a Poisson\n",
    "    Log-Normal count model. This allows\n",
    "    the total read depth to vary across\n",
    "    samples which can be modeled by the\n",
    "    dispersion perameter of the Poisson.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    table : array like\n",
    "        A matrix of counts.\n",
    "    depth : int\n",
    "        Mean read depth of the\n",
    "        simulation.\n",
    "    read_std : float\n",
    "        Controls dispersion.\n",
    "    pnormal : float [0,1]\n",
    "        Percent\n",
    "        normally-distributed\n",
    "        noise to the data.\n",
    "    pheter : float [0,1]\n",
    "        Percent random\n",
    "        noise to the data.\n",
    "    pzero : float [0,1]\n",
    "        Percent zeros added.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    sim : array-like\n",
    "        simulated counts.\n",
    "\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Äijö, Tarmo, Christian L. Müller,\n",
    "           and Richard Bonneau. 2018.\n",
    "           “Temporal Probabilistic Modeling of Bacterial\n",
    "           Compositions Derived from 16S rRNA Sequencing.”\n",
    "           Bioinformatics  34 (3): 372–80.\n",
    "    \"\"\"\n",
    "\n",
    "    # subset and order single digester\n",
    "    sim = X.copy()\n",
    "    # (1) estimate proportions\n",
    "    sim = closure(sim)\n",
    "    # (2) filter low-mean reads\n",
    "    abund_filt_mask = sim.mean(axis=0) > 1e-5\n",
    "    mask_index = [i for i, v in enumerate(abund_filt_mask) if v]\n",
    "    sim = sim.T[abund_filt_mask].T\n",
    "    # (3) median filter by feature\n",
    "    if mwindow:\n",
    "        for i in range(sim.shape[1]):\n",
    "            sim[:, i] = medfilt(sim[:, i], mwindow)\n",
    "    sim = closure(sim)  # check closure\n",
    "    # (4) project onto real space\n",
    "    # with inverse softmax\n",
    "    sim = softmax_inv(sim)\n",
    "    # (5) noise gauss (@10% of samps.)\n",
    "    err = pnormal * np.ones_like(sim)\n",
    "    sim = normal(sim, err)\n",
    "    # add some random noise (@10% of samps.)\n",
    "    err = pheter * np.ones_like(sim)\n",
    "    i = randint(0, err.shape[0], 5000)\n",
    "    j = randint(0, err.shape[1], 5000)\n",
    "    err[i, j] = 0.5\n",
    "    sim = normal(sim, err)\n",
    "    # add random zeros (@10% of samps.)\n",
    "    err = pzero * np.ones_like(sim)\n",
    "    i = randint(0, err.shape[0], 100)\n",
    "    j = randint(0, err.shape[1], 100)\n",
    "    err[i, j] = 0.0\n",
    "    sim = normal(sim, err)\n",
    "    # (6) back to prop\n",
    "    sim = closure(softmax(sim))\n",
    "    # (7) Poisson - Log-Normal Counts\n",
    "    mean = depth\n",
    "    stdev = read_std\n",
    "    phi = (stdev ** 2 + mean ** 2) ** 0.5\n",
    "    mu = mean**2 / phi\n",
    "    mu = np.log(mu * sim.T)\n",
    "    sigma = (np.log(phi ** 2 / mean ** 2)) ** 0.5\n",
    "    sim = np.vstack([poisson(lognormal(mu[:, i],\n",
    "                                       sigma))\n",
    "                     for i in range(sim.shape[0])])\n",
    "    return sim, mask_index, sigma\n",
    "\n",
    "def simulation_helper(bt, mf, \n",
    "                      subject_id_column,\n",
    "                      state_column,\n",
    "                      depth=100, read_std=10.0,\n",
    "                      pnormal=0.1, pheter=0.1,\n",
    "                      pzero=0.1, mwindow=False):\n",
    "    \"\"\"\n",
    "    helper function to run simulation multiple times.\n",
    "    \"\"\"\n",
    "    # full simulation table\n",
    "    simulation_table = []\n",
    "    # generate simulation by series\n",
    "    table_df = pd.DataFrame(bt.matrix_data.toarray(), bt.ids('observation'), bt.ids())\n",
    "    for subject_id, subject_mf in mf.groupby(subject_id_column):\n",
    "        # subset to single series\n",
    "        subject_mf = subject_mf.sort_values(state_column)\n",
    "        table_df_ = table_df.copy().loc[:, subject_mf.index]\n",
    "        table_df_ = table_df_[table_df_.sum(1) > 0].T\n",
    "        # continue\n",
    "        if table_df_.shape[0] <= 1:\n",
    "            continue\n",
    "        # generate simulation on series \n",
    "        # This is where you change data characteristics!!\n",
    "        sim, mask_index, sigma = simulate(table_df_.values,\n",
    "                                          depth=depth,\n",
    "                                          read_std=read_std,\n",
    "                                          pnormal=pnormal,\n",
    "                                          pheter=pheter,\n",
    "                                          pzero=pzero,\n",
    "                                          mwindow=mwindow)\n",
    "        # make dataframe\n",
    "        simdf = pd.DataFrame(sim.T,\n",
    "                             table_df_.columns[mask_index],\n",
    "                             table_df_.index)\n",
    "        simulation_table.append(simdf)\n",
    "    # merge\n",
    "    simulation_table = pd.concat(simulation_table, axis=1, sort=False).fillna(0)\n",
    "    # convert table to biom\n",
    "    simulation_table = Table(simulation_table.values, \n",
    "                             simulation_table.index, \n",
    "                             simulation_table.columns)\n",
    "    return simulation_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load filtered tables\n",
    "body_site = 'Baby-Feces'\n",
    "mf = pd.read_csv('../data/%s/metadata-filtered.tsv' % body_site, index_col=0)\n",
    "bt = load_table('../data/%s/table-filtered.biom' % body_site)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make sure there are at least two time points per subject\n",
    "subject_id_column = 'subjectid_unique'\n",
    "ids_to_keep = mf[subject_id_column].value_counts()[mf[subject_id_column].value_counts() > 1].index\n",
    "mf = mf[mf[subject_id_column].isin(ids_to_keep)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inits\n",
    "experiment = 'seq-depth'\n",
    "results_path = '../results/{}-tables/'.format(experiment)\n",
    "state_column =  'date_sampling_category_days_continuous'\n",
    "simulations = {}\n",
    "# example: altering the depth example\n",
    "for exp in tqdm([500, 1000, 5000, 10000]):\n",
    "    for fold in range(3):\n",
    "        dispersion = int(exp/10)\n",
    "        print('Depth: {}, fold: {}'.format(exp, fold))\n",
    "        simulations[(dispersion, fold)] = simulation_helper(bt, mf, subject_id_column, state_column,\n",
    "                                                            depth=exp, read_std=dispersion)\n",
    "        \n",
    "        with biom_open('{}table-{}-{}.biom'.format(results_path, exp, fold), 'w') as f:  \n",
    "            simulations[(dispersion, fold)].to_hdf5(f, \"{}-{}\".format(dispersion, fold))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1754c651ef11442e85118cedbaaf06de",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dispersion: 200.0, fold: 0\n",
      "Dispersion: 200.0, fold: 1\n",
      "Dispersion: 200.0, fold: 2\n",
      "Dispersion: 100.0, fold: 0\n",
      "Dispersion: 100.0, fold: 1\n",
      "Dispersion: 100.0, fold: 2\n",
      "Dispersion: 50.0, fold: 0\n",
      "Dispersion: 50.0, fold: 1\n",
      "Dispersion: 50.0, fold: 2\n",
      "Dispersion: 20.0, fold: 0\n",
      "Dispersion: 20.0, fold: 1\n",
      "Dispersion: 20.0, fold: 2\n"
     ]
    }
   ],
   "source": [
    "# inits\n",
    "experiment = 'dispersion'\n",
    "results_path = '../results/{}-tables/'.format(experiment)\n",
    "state_column =  'date_sampling_category_days_continuous'\n",
    "simulations = {}\n",
    "# example: altering the depth example\n",
    "for exp in tqdm([5,10,20,50]):\n",
    "    for fold in range(3):\n",
    "        seq_depth = 1000\n",
    "        dispersion = int(seq_depth/exp)\n",
    "        print('Dispersion: {}, fold: {}'.format(seq_depth/exp, fold))\n",
    "        simulations[(dispersion, fold)] = simulation_helper(bt, mf, subject_id_column, state_column,\n",
    "                                                            depth=seq_depth, read_std=dispersion)\n",
    "        \n",
    "        with biom_open('{}table-{}-{}-{}.biom'.format(results_path, seq_depth, dispersion, fold), 'w') as f:  \n",
    "            simulations[(dispersion, fold)].to_hdf5(f, \"{}-{}\".format(dispersion, fold))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-amplicon-2024.2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
